### DISCUSSION TITLE
Schema definition circular references

### DISCUSSION THREAD

**Discussion by EmileSpecs** [feathersjs/feathers#3473] [Q&A]
Hi

Please see https://github.com/sinclairzx81/typebox/issues/844 for reference.

The solution offered unfortunately won't work in the context of feathers since none of the other functions like the `TReference` type produced for schema.

For example `getValidator` wants schema:
`schema: TObject<TProperties> | TIntersect<TObject<TProperties>[]> | TUnion<TObject<TProperties>[]>`

Is there any other suggested way of defining references to other schemas that reference each other that will work?

Reference to my definitions as mentioned in the post above:
```
// NOTE: these schemas are defined in different files and are displayed here as is for reference, thus I'm not getting errors like:
// Block-scoped variable 'accountSchema' used before its declaration.ts

const userSchema = Type.Object(
  {
    id: Type.String({ format: 'uuid' }),
    email: Type.String({ format: 'email' }),
    account: Type.Ref(accountSchema)
  },
  { $id: 'User', additionalProperties: false }
)

const accountSchema = Type.Object(
  {
    id: Type.String({ format: 'uuid' }),
    name: Type.String(),
    users: Type.Array(Type.Ref(userSchema)),
    locations: Type.Array(Type.Ref(locationSchema))
  },
  { $id: 'Account', additionalProperties: false }
)
```

The provided solution that does not solve the issue for me in feathers:
```
const userSchemaBuilder = <T extends TSchema>(reference: T) =>
  Type.Object({
    id: Type.String({ format: "uuid" }),
    email: Type.String({ format: "email" }),
    account: Type.Optional(reference),
  });

const accountSchemaBuilder = <T extends TSchema>(reference: T) =>
  Type.Object({
    id: Type.String({ format: "uuid" }),
    name: Type.String(),
    users: Type.Optional(Type.Array(reference)),
  });

const userSchema = Type.Recursive((This) =>
  userSchemaBuilder(accountSchemaBuilder(This))
);
const accountSchema = Type.Recursive((This) =>
  accountSchemaBuilder(userSchemaBuilder(This))
);
type User = Static<typeof userSchema>;
type Account = Static<typeof accountSchema>;
```

Would appreciate some help on how to make this work!

**Comment by EmileSpecs** [comment]
It seems to me like the most simple solution here is just to use JSON Schema instead, as it doesn't give any errors when doing the same circular relations.

Any opinions, as I know TypeBox is the preferred option?
Doesn't TypeBox just compile to JSON Schema anyway?

**Comment by daffl** [comment]
You can always fall back fall back to JSON schema directly with [Type.Unsafe](https://github.com/sinclairzx81/typebox?tab=readme-ov-file#unsafe-types).

```ts
const userSchema = Type.Object(
  {
    id: Type.String({ format: 'uuid' }),
    email: Type.String({ format: 'email' }),
    account: Type.Ref(accountSchema)
  },
  { $id: 'User', additionalProperties: false }
)

const accountSchema = Type.Object(
  {
    id: Type.String({ format: 'uuid' }),
    name: Type.String(),
    users: Type.Array(Type.Unsafe<User>({ $ref: 'User' })),
    locations: Type.Array(Type.Ref(locationSchema))
  },
  { $id: 'Account', additionalProperties: false }
)
```

**Reply by EmileSpecs** [reply]
Thanks @daffl!

Seemed like a good solution but unfortunately the error remains.

I get an additional error for `type User = Static<typeof userSchema>`: `Type alias 'User' circularly references itself.`
The original error also remains.

I also changed the account reference in the user schema to the same formatting then I get the circular reference error on both ends.

Probably just something small that's missing?

**Reply by daffl** [reply]
Oh true, circular type inference is not actually possible with TypeScript (this will also not work with JSON schema). You might have to stick with `Type.Unsafe<any>` and cast it to the correct type when you use it.

**Reply by EmileSpecs** [reply]
Actually this does seem to work fine in JSON Schema:

```
export const subscriptionSchema = {
  $id: 'Subscription',
  type: 'object',
  additionalProperties: false,
  required: ['id', 'text'],
  properties: {
    id: { type: 'number' },
    text: { type: 'string' },
    // Schema references
    invoices: { type: 'array', items: { $ref: 'Invoice' }}
  }
} as const
export type Subscription = FromSchema<typeof subscriptionSchema,
{
  // All schema references need to be passed to get the correct type
  references: [typeof invoiceSchema]
}>

export const invoiceSchema = {
  $id: 'Invoice',
  type: 'object',
  additionalProperties: false,
  required: ['id', 'text'],
  properties: {
    id: { type: 'number' },
    text: { type: 'string' },
    // Schema references
    subscription: { $ref: 'Subscription' }
  }
} as const
export type Invoice = FromSchema<typeof invoiceSchema,
{
  // All schema references need to be passed to get the correct type
  references: [typeof subscriptionSchema]
}>
```

This example produces no errors and seems to work as expected. Interestingly it doesn't map the reference recursively as the `subscription` for example is typed as `undefined` once it gets to that second level of recursion which is perfect:
```
type Subscription = {
    invoices?: {
        subscription?: undefined;
        id: number;
        text: string;
    }[] | undefined;
    id: number;
    text: string;
}
```

So unless I'm missing something this does seem to work and I guess I'll just be using JSON Schema rather than TypeBox as there are many such relationships where I would like to populate data from both sides of the relationship.

What would you do @daffl? Would you also opt for JSON Schema or just use `Type.Unsafe<any>`? I don't know enough to weigh in on the pros and cons of each method.

**Reply by EmileSpecs** [reply]
Nevermind, seems like everything worked fine until I started resolving those properties.
I then got an odd error about the type being declared twice...?

So I've now gone the `Type.Unsafe<any>` route and that seems to work fine.

**Reply by daffl** [reply]
Interesting that it worked so far with JSON schema but in general, I found TypeScript doing really weird things with circular types which is why I try to avoid it.

**Reply by EmileSpecs** [reply]
The perhaps expected end result of all of this is that I now have an endless resolver loop.

```
export const userResolver = resolve<User, HookContext<UserService>>({
  // THIS RESOLVER EXECUTES IN ENDLESS LOOP
  account: virtual(async (user, context) => {
    // Populate the account field with the actual account data
    if (!user.account_id) {
      return
    }
    return await context.app.service('accounts').get(user.account_id)
  })
})
```

If this was a normal hook I would add a property to `context.params` to indicate that the user has been resolved but in the resolver the context remains static.

Any way to solve this in the property resolver or do I now have to use a normal hook?

**Reply by daffl** [reply]
I thought I documented this but couldn't find it 🤔. There should be a guard against recursive resolving when passing `params.resolve` along:

```ts
export const userResolver = resolve<User, HookContext<UserService>>({
  // THIS RESOLVER EXECUTES IN ENDLESS LOOP
  account: virtual(async (user, context) => {
    // Populate the account field with the actual account data
    if (!user.account_id) {
      return
    }
    const { resolve } = context.params
    return await context.app.service('accounts').get(user.account_id, { resolve })
  })
})
```

**Reply by EmileSpecs** [reply]
Hi @daffl 

Unfortunately that didn't resolve it, it still endlessly loops that property resolver.
Bug?

**Reply by EmileSpecs** [reply]
I also added the resolve parameter to the account schema side:
```
export const accountResolver = resolve<Account, HookContext<AccountService>>({
  users: virtual(async (account, context) => {
    // Populate the users field with the actual user data
    const { resolve } = context.params
    const users = await context.app.service('users').find({ query: { account_id: account.id }, paginate: false, resolve })

    return users?.length ? users : undefined
  })
})
```
That made no difference either.

**Reply by daffl** [reply]
Could be a bug but it'd be easier to verify with a minimal example to reproduce.

**Reply by EmileSpecs** [reply]
Here you go!

https://github.com/EmileSpecs/f5-resolve

Get it set up and then just query `users` or `accounts` and watch the console go nuts.
